src/context.ts

```diff
@@ -11,10 +11,10 @@ import { EqualityChecker, State, StateSelector, UseBoundStore } from 'zustand'
 /**
  * @deprecated Use `typeof MyContext.useStore` instead.
  */
-export type UseContextStore<T extends State> = <U = T>(
-  selector?: StateSelector<T, U>,
-  equalityFn?: EqualityChecker<U>
-) => U
+export type UseContextStore<T extends State> = {
+  (): T
+  <U>(selector: StateSelector<T, U>, equalityFn?: EqualityChecker<U>): U
+}
```

Comment:
@dai-shi Shouldn't this have been the following?

```ts
export type UseContextStore<T extends State> = {
  (): T
  // note the "?" after selector
  <U>(selector?: StateSelector<T, U>, equalityFn?: EqualityChecker<U>): U
}
```

As `useContextStore(undefined, myEquals)` is also valid.

Should I correct it in #725 or make a new PR?

Reply:
That causes the DX issue in #812.

https://tsplay.dev/WJq5Vm

We don't recommend/allow passing `undefined` for selector as a practice. So, think it like TS types are correct, and JS impl is just naive. (We could use `arguments`, but it's not worth the size&complexity.)

btw, this is the similar discussion for `combine` in https://github.com/pmndrs/zustand/pull/725#discussion_r845632251. TS types are correct for usage, so no need to be too precise about typing JS impl (if it's not the recommended usage. Could be done with `defineProperty`, but...)

Reply:
That's trivial to fix, we can just add another overload — https://tsplay.dev/WGPAJw — we already have 2 overloads so adding another doesn't make a difference.

I think not supporting `undefined` selector is an unnecessary compromise, what if users want to actually use custom equality function without selector then they'd have to do `useStore(x => x, eq)` which looks less expressive and less intuitive than `useStore(undefined, eq)`.

So I insist that we add another overload, I'm doing it in #725, if you don't like it I'll revert.

And in case of `combine` we're being intentionally unprecise with types to make the inference work, just that I think it's important to inform the users about the unsoundness we have produced, even if it is negligible. (Also shallow merge is good, `defineProperty` would make it more confusing and won't even solve the problem.)

Reply:
I don't like `useStore(undefined, eq)`, and it's almost always a mistake in the usage. But, having another overload is acceptable for an edge case. Will look at #725 and comment there.